zio-http/src/main/scala/zio/http/endpoint/internal/EndpointClient.scala

```diff
@@ -41,15 +41,47 @@ private[endpoint] final case class EndpointClient[I, E, O, M <: EndpointMiddlewa
         // Preferentially decode an error from the handler, before falling back
         // to decoding the middleware error:
         val leftError =
-          endpoint.error.decodeResponse(response).orDie.flatMap((e: E) => ZIO.fail(alt.left(e)))
+          endpoint.error.decodeResponse(response).map(e => alt.left(e))
 
         val rightError =
           invocation.middleware.error
             .decodeResponse(response)
-            .orDie
-            .flatMap((e: invocation.middleware.Err) => ZIO.fail(alt.right(e)))
+            .map(e => alt.right(e))
 
-        leftError.orElse(rightError)
+        def generateErrorMessage(t1: Throwable, t2: Throwable) = {
+          Cause.Both(
+            Cause
+              .fail(
+                new IllegalStateException(s"Failed to deserialize response. Response status: ${response.status}"),
+              ),
+            Cause.Both(
+              Cause
+                .die(new IllegalStateException("Cannot deserialize using endpoint error codec", t1)),
+              Cause
+                .die(new IllegalStateException("Cannot deserialize using middleware error codec", t2)),
+            ),
+          )
+        }
+
+        def orElseCombineErrors[R, E1, A](
+          zio1: ZIO[R, E1, A],
+          zio2: ZIO[R, E1, A],
+          f: (E1, E1) => Cause[E1],
+        ): ZIO[R, E1, A] = {
+          zio1.foldZIO(
+            success = ZIO.succeed(_),
+            failure = cause1 => {
+              zio2.foldZIO(
+                success = ZIO.succeed(_),
+                failure = cause2 => {
+                  ZIO.failCause(f(cause1, cause2))
+                },
+              )
+            },
+          )
+        }
+
+        orElseCombineErrors(leftError, rightError, generateErrorMessage).orDie.flip
```

Comment:
You could just combine the two failure like this:

```scala
first.catchAllCause { firstCause =>
  second.catchAllCause { secondCause =>
    ZIO.failCause(secondCause ++ firstCause)
  }
}
```

and attach the additional information (those messages above) via `mapError` to each level (`first.mapError`, `second.mapError`)